akka-actor/src/main/scala/akka/util/MessageBuffer.scala

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import akka.actor.ActorRef
+
+/**
+ * A single producer single consumer mutable message buffer
+ */
+final class MessageBuffer private (
+  private var _head: MessageBuffer.Node,
+  private var _tail: MessageBuffer.Node) {
```

Comment:
Yes I know that but on the other hand appending and spilling over the capacity of the `ArrayDeque` will result in a full array copy of all the elements to the new bigger array.

I focused on making the `foreach` not allocate a tuple for every element. I think that will be the biggest win here.